precompile/allowlist/role.go

```diff
@@ -5,79 +5,80 @@ package allowlist
 
 import (
 	"errors"
+	"fmt"
 	"math/big"
 
 	"github.com/ethereum/go-ethereum/common"
+	"github.com/holiman/uint256"
 )
 
-// 1. NoRole - this is equivalent to common.Hash{} and deletes the key from the DB when set
-// 2. EnabledRole - allowed to call the precompile
-// 3. Admin - allowed to both modify the allowlist and call the precompile
-// 4. Manager - allowed to add and remove only enabled addresses and also call the precompile. (only after Durango)
-var (
-	NoRole      = Role(common.BigToHash(common.Big0))
-	EnabledRole = Role(common.BigToHash(common.Big1))
-	AdminRole   = Role(common.BigToHash(common.Big2))
-	ManagerRole = Role(common.BigToHash(common.Big3))
-	// Roles should be incremented and not changed.
-
-	ErrInvalidRole = errors.New("invalid role")
+// Role mirrors the Solidity enum of the same name.
+type Role uint64
+
+// Role constants; they MUST be incremented and MUST NOT be reordered.
+const (
+	// NoRole - this is equivalent to common.Hash{} and deletes the key from the DB when set
+	NoRole Role = iota
+	// EnabledRole - allowed to call the precompile
+	EnabledRole
+	// AdminRole - allowed to both modify the allowlist and call the precompile
+	AdminRole
+	// ManagerRole - allowed to add and remove only enabled addresses and also call the precompile. (only after Durango)
+	ManagerRole
+
+	// firstInvalidRole is useful for loops and other upper-bound checks of
+	// Role values. See [Role.IsValid].
+	firstInvalidRole
 )
 
-// Enum constants for valid Role
-type Role common.Hash
+// ErrInvalidRole is returned if an invalid role is encountered. If this error
+// is returned along with a [Role] value, said value MUST NOT be used; this
+// allows functions to return `Role(0), ErrInvalidRole` without worrying about
+// misinterpretation as [NoRole].
+var ErrInvalidRole = errors.New("invalid role")
 
-// IsNoRole returns true if [r] indicates no specific role.
-func (r Role) IsNoRole() bool {
-	switch r {
-	case NoRole:
-		return true
-	default:
-		return false
-	}
-}
-
-// IsAdmin returns true if [r] indicates the permission to modify the allow list.
-func (r Role) IsAdmin() bool {
-	switch r {
-	case AdminRole:
-		return true
-	default:
-		return false
-	}
+// IsValid returns whether or not `r` is a valid value.
```

Comment:
nit: while ` is more broadly used avalanche code bases use [] for this type of notation.